File: packages/app/src/services/software/definitions/darktable.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory } from 'shared';
import { GithubReleasesScraper } from '../githubReleasesScraper';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'darktable';
	public name = 'darktable';
	public category = [SoftwareCategory.Creative, SoftwareCategory.Media];
	public downloadName = 'darktable-win64.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public deprecated = false;
	public icon = 'darktable.png';
	public homepage = 'https://darktable.org';

	public async resolveDownloadUrl() {
		const owner   = 'darktable-org';
		const repo    = 'darktable';
		const release = await GithubReleasesScraper.getRelease(owner, repo, r => r.endsWith('-win64.exe'));
		if (release.isErr()) {
			return err(release.error);
		}

		return ok(release.value);
	}
}

File: src/lib/software/darktable.py
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from src.lib.github_release_scraper import GithubReleaseScraper

class Darktable(BaseSoftware):
    def __init__(self):
        super().__init__()

        self._gh = GithubReleaseScraper('darktable-org', 'darktable', self)
        self._gh.releases_scraped.connect(self._on_releases_scraped)

        self.key = 'darktable'
        self.name = 'darktable'
        self.category = [SoftwareCategory.Creative, SoftwareCategory.Media]
        self.download_name = 'darktable-win64.exe'
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'darktable.png'
        self.homepage = 'https://darktable.org'

    @Slot(list)
    def _on_releases_scraped(self, releases: list[str]):
        asset = next((release for release in releases if release.endswith('win64.exe')), None)
        if asset:
            self.url_resolved.emit(asset)
        else:
            self.url_resolve_error.emit(self.ResolveError.GitHubAssetNotFoundError)

    def resolve_download_url(self):
        self._gh.get_repo_releases()
